test/functional/apps/discover/group3/_request_counts.ts

```diff
@@ -183,7 +183,7 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
         });
       });
 
-      it('should send 2 requests (documents + chart) when adding a filter', async () => {
+      it.skip('should send 2 requests (documents + chart) when adding a filter', async () => {
```

Comment:
Still looking into that. What concerns is that adding a filter is a common action for Discover and if it triggers unnecessary requests then we should investigate more before skipping and merging.

Reply:
I think the counter is sometimes 3 instead of 2 because there is a cancelled request in between. Here is an example:

```
{
  "connectEnd": 12212.799999982119,
  "connectStart": 12212.799999982119,
  "decodedBodySize": 3616665,
  "deliveryType": "",
  "domainLookupEnd": 12212.799999982119,
  "domainLookupStart": 12212.799999982119,
  "duration": 229.30000001192093,
  "encodedBodySize": 393756,
  "entryType": "resource",
  "fetchStart": 12212.799999982119,
  "firstInterimResponseStart": 0,
  "initiatorType": "fetch",
  "name": "http://localhost:5620/internal/search/ese",
  "nextHopProtocol": "http/1.1",
  "redirectEnd": 0,
  "redirectStart": 0,
  "renderBlockingStatus": "non-blocking",
  "requestStart": 12213.299999982119,
  "responseEnd": 12442.09999999404,
  "responseStart": 12405.899999976158,
  "responseStatus": 200,
  "secureConnectionStart": 0,
  "serverTiming": [],
  "startTime": 12212.799999982119,
  "toJSON": {},
  "transferSize": 394056,
  "workerStart": 0
}
{
  "connectEnd": 0,
  "connectStart": 0,
  "decodedBodySize": 0,
  "deliveryType": "",
  "domainLookupEnd": 0,
  "domainLookupStart": 0,
  "duration": 100.30000001192093,
  "encodedBodySize": 0,
  "entryType": "resource",
  "fetchStart": 12263.899999976158,
  "firstInterimResponseStart": 0,
  "initiatorType": "fetch",
  "name": "http://localhost:5620/internal/search/ese",
  "nextHopProtocol": "",
  "redirectEnd": 0,
  "redirectStart": 0,
  "renderBlockingStatus": "non-blocking",
  "requestStart": 0,
  "responseEnd": 12364.199999988079,
  "responseStart": 0,
  "responseStatus": 0,
  "secureConnectionStart": 0,
  "serverTiming": [],
  "startTime": 12263.899999976158,
  "toJSON": {},
  "transferSize": 0,
  "workerStart": 0
}
{
  "connectEnd": 12289,
  "connectStart": 12289,
  "decodedBodySize": 2308,
  "deliveryType": "",
  "domainLookupEnd": 12289,
  "domainLookupStart": 12289,
  "duration": 15.099999994039536,
  "encodedBodySize": 524,
  "entryType": "resource",
  "fetchStart": 12289,
  "firstInterimResponseStart": 0,
  "initiatorType": "fetch",
  "name": "http://localhost:5620/internal/search/ese",
  "nextHopProtocol": "http/1.1",
  "redirectEnd": 0,
  "redirectStart": 0,
  "renderBlockingStatus": "non-blocking",
  "requestStart": 12289.399999976158,
  "responseEnd": 12304.09999999404,
  "responseStart": 12303.799999982119,
  "responseStatus": 200,
  "secureConnectionStart": 0,
  "serverTiming": [],
  "startTime": 12289,
  "toJSON": {},
  "transferSize": 824,
  "workerStart": 0
}
```

What if adjust 
```
const getSearchCount = async (type: 'ese' | 'esql') => {
  const requests = await browser.execute(() =>
    performance
      .getEntries()
      .filter((entry: any) => ['fetch', 'xmlhttprequest'].includes(entry.initiatorType))
  );
  return requests.filter((entry) => entry.name.endsWith(`/internal/search/${type}`)).length;
};
```

to the following to filter out canceled requests?

```
const getSearchCount = async (type: 'ese' | 'esql') => {
  const requests = await browser.execute(() =>
    performance
      .getEntries()
      .filter((entry: any) => ['fetch', 'xmlhttprequest'].includes(entry.initiatorType))
  );
  return requests.filter((entry) => entry.name.endsWith(`/internal/search/${type}`) && ('responseStatus' in entry && entry. responseStatus)).length;
};
```

Reply:
Looks like it makes a request with the previous time range and then immediately makes a request with a new time range.

Reply:
Not sure why it was not surfacing before but looks like the cause for such issues is that `filters` and `timeRange` props update at different moments of time. 

https://github.com/dej611/kibana/blob/722a913c54cd1521d615f9b093d6dd495c65fde9/src/plugins/discover/public/application/main/components/layout/use_discover_histogram.ts#L388-L390

Ideally we could debounce the updates and provide all new props (query, filters, timeRange) to UnifiedHistogram at once.

@dej611 Do you know why only timeRange gets passed and updated in `useSearchApi` hook and not filters? How is it being used?

https://github.com/dej611/kibana/blob/b6287708f687d4e3288851052c0c6ae4ade8ce60/packages/presentation/presentation_publishing/interfaces/fetch/publishes_unified_search.ts#L123

Is it correct that `timeRange` gets passed directly as props and also inside `searchApi`:

https://github.com/dej611/kibana/blob/aaf1f3107ea87a45bef4441e13dd3982cb1cf895/x-pack/plugins/lens/public/react_embeddable/renderer/lens_custom_renderer_component.tsx#L85-L88

Reply:
From what I observed only `props.timeRange` was updating. `props.filters` did not.

Reply:
Checked again locally and now it seems to pass. I've restored it back, let's see how it goes.

Reply:
No luck, still failing while it's passing locally. It must be some race condition.
I'll try to investigate this on Monday again.

Reply:
For reproducing this issue I was adding the filter but also removing it. After the removal I often saw 3 requests.

Reply:
I looked into this too and was able to reproduce the same behaviour.

I don't think it's related to the `filters` and `timeRange` changing at different times in `use_discover_histogram` because we avoid updating the actual props we pass to Lens from within Unified Histogram until its `refetch$` observable is triggered, specifically to avoid additional Lens fetches when Unified Histogram props change: https://github.com/elastic/kibana/blob/2296a791790bd4fa0b5dd5d22181aa3549af941c/src/plugins/unified_histogram/public/chart/hooks/use_lens_props.ts#L49-L58

I put a breakpoint inside `updateLensPropsContext` and confirmed that it was only triggered once while reproducing the issue, meaning Lens only got prop updates once.

It looks like the issue is lower down in `x-pack/plugins/lens/public/react_embeddable/data_loader.ts`. Whenever I added or removed a filter, the `reload` function was triggered twice -- once with `attributes` and once with `searchContext`.

It triggers once due to the `attributes` prop change here: https://github.com/elastic/kibana/blob/06c715d854b8a3a5529267e7ef4801ab335b78ef/x-pack/plugins/lens/public/react_embeddable/renderer/lens_custom_renderer_component.tsx#L57-L66

And then again due to to the `timeRange` prop change here: https://github.com/elastic/kibana/blob/06c715d854b8a3a5529267e7ef4801ab335b78ef/packages/presentation/presentation_publishing/interfaces/fetch/publishes_unified_search.ts#L122-L124

We do pass both the `attributes` and `timeRange` props from Unified Histogram, but they're both updated at the same time. It looks like it just triggers `reload` twice due to the separate `useEffect` calls.

One thing I tried locally that seemed to fix it was adding a small debounce to the `reload` function so it only fires once when multiple observables are triggered in succession. After that it consistently fired a single request and only logged `[Lens] Embeddable reload reason` once with `window.ELASTIC_LENS_LOGGER = true`. I'm not sure if that would have a negative impact outside of Discover though.

On a side note, a lot of this complexity comes down to the auto fetching behaviour of the Lens embeddable. I feel like the new embeddable architecture could make it easier to disable auto fetching and allow consumers to choose when fetches occur if they wanted to (not in this PR of course). This would allow us to greatly simplify the code around it in Discover and Unified Histogram.

Reply:
I think I've found the main issue here.
With [5560171](https://github.com/elastic/kibana/pull/186642/commits/5560171fef056488ac02335a06e9234cef827bcb) I'm merging the observables from both `fetch` and other attribute changes with a `debounceTime` to increase the probability to batch more and more.
The result of that was actually a slight improvement on the request count, with scenarios that can have even less requests than what expected. It depends on the `debounce` so it's not deterministic and I had to adapt the test to handle it (using the current count as upper bound, so it can get only better, no worst).

I did some manual check and it seems to work ok, but will wait the CI outcome to verify how effective it is.